nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/StreamBulkScroller.java

```diff
@@ -0,0 +1,173 @@
+            this.scrollKeepAliveSeconds = scrollKeepAliveSeconds;
+        }
+
+        @Override
+        public void processRecord(ComputationContext context, String inputStreamName, Record record) {
+            TransactionHelper.runInTransaction(() -> processRecord(context, record));
+        }
+
+        protected void processRecord(ComputationContext context, Record record) {
+            try {
+                BulkCommand command = getBulkCommandJson(record.data);
+                if (!kvStore.compareAndSet(record.key + STATE, SCHEDULED.toString(), BUILDING.toString())) {
+                    log.error("Discard record: " + record + " because it's already building");
+                    return;
+                }
+                try (CloseableCoreSession session = CoreInstance.openCoreSession(command.getRepository(),
+                        command.getUsername())) {
+                    // scroll documents
+                    ScrollResult<String> scroll = session.scroll(command.getQuery(), scrollBatchSize,
+                            scrollKeepAliveSeconds);
+                    long documentCount = 0;
+                    while (scroll.hasResults()) {
+                        List<String> docIds = scroll.getResults();
+                        // send these ids as keys to the appropriate stream
+                        // key will be bulkOperationId/docId
+                        // value/data is a BulkCommand serialized as JSON
+                        docIds.forEach(docId -> context.produceRecord(command.getOperation(), record.key + '/' + docId,
+                                record.data));
+                        documentCount += docIds.size();
+                        // next batch
+                        scroll = session.scroll(scroll.getScrollId());
+                        TransactionHelper.commitOrRollbackTransaction();
+                        TransactionHelper.startTransaction();
+                    }
+                    context.askForCheckpoint();
+                    kvStore.put(record.key + STATE, COMPLETED.toString());
+                    kvStore.put(record.key + SCROLLED_DOCUMENT_COUNT, documentCount);
+                }
+            } catch (NuxeoException e) {
+                log.error("Discard invalid record: " + record, e);
+            }
+        }
+
+        protected BulkCommand getBulkCommandJson(byte[] data) {
+            String json = "";
+            try {
+                json = new String(data, UTF_8);
+                ObjectMapper mapper = new ObjectMapper();
+                return mapper.readValue(json, BulkCommand.class);
+            } catch (UnsupportedEncodingException e) {
+                throw new NuxeoException("Discard bulk command, invalid byte array", e);
+            } catch (IOException e) {
+                throw new NuxeoException("Invalid json bulkCommand=" + json, e);
+            }
+        }
+    }
+
+    // TODO copied from StreamAuditWriter - where can we put that ?
```

Comment:
we might want to add an AbstractStreamProcessorTopology with helper

Reply:
Forgot this one, we though about abstract, default method in interface or static method in interface. 3 possibilities, 3 devs, no solutions...
Better is abstract ?